File: test/util_tests.cpp
#include <chrono>
#include <iostream>

#include <fmt/core.h>
#include <fmt/chrono.h>

#include <gtest/gtest.h>

#include "Util.h"
#include "error/Exceptions.h"


TEST(BasicTests, ParseURLParameterTest) {

    std::string url = "test.local/tenants/XXX/connectors/XXX/devices?page[size]=1&page[before]=Mzk5ZTM1MWYtNTI3OS00YzFhLTk0MmUtYTZiODBmMjFiYzVh";

    std::map<std::string, std::string> params = BASCloud::Util::parseURLParameter(url);

    EXPECT_EQ(params.at("page[size]"), "1");
    EXPECT_EQ(params.at("page[before]"), "Mzk5ZTM1MWYtNTI3OS00YzFhLTk0MmUtYTZiODBmMjFiYzVh");
}

TEST(BasicTests, ParseCurrentDateTimeTest) {

    std::time_t currentDateTime = std::chrono::duration_cast<std::chrono::seconds>(std::chrono::system_clock::now().time_since_epoch()).count();

    std::cerr <<"CURRENT DATETIME: "<<currentDateTime<<std::endl;

    // "2020-03-09T13:09:41.640Z"
    std::string currentDateTimeString = fmt::format("{:%FT%T.000Z}", fmt::localtime(currentDateTime));

    std::cerr <<"CURRENT DATETIME: "<<currentDateTimeString<<std::endl;

    std::time_t parsedDateTime = BASCloud::Util::parseDateTimeString(currentDateTimeString);
    std::string parsedDateTimeString = fmt::format("{:%FT%T.000Z}", fmt::localtime(parsedDateTime));

    std::cerr <<"PARSED DATETIME: "<<parsedDateTime<<std::endl;
    std::cerr <<"PARSED DATETIME: "<<parsedDateTimeString<<std::endl;

    EXPECT_EQ(currentDateTime, parsedDateTime);
    EXPECT_EQ(currentDateTimeString, parsedDateTimeString);
}

TEST(BasicTests, ParseFixedDateTimeTest) {

    std::string fixedDateTimeString = "2019-06-02T12:17:58.819Z";

    std::cerr <<"FIXED DATETIME: "<<fixedDateTimeString<<std::endl;

    std::time_t parsedDateTime = BASCloud::Util::parseDateTimeString(fixedDateTimeString);
    std::string parsedDateTimeString = fmt::format("{:%FT%T.819Z}", fmt::localtime(parsedDateTime));

    std::cerr <<"PARSED DATETIME: "<<parsedDateTime<<std::endl;
    std::cerr <<"PARSED DATETIME: "<<parsedDateTimeString<<std::endl;

    EXPECT_EQ(fixedDateTimeString, parsedDateTimeString);
}

int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: pyBAScloudAPI/tests/util_tests.py
import unittest
import time 
import datetime

import pyBAScloudAPI as api


class TestParseURLParameter(unittest.TestCase):

    def test_parsing(self):
        url = "test.local/tenants/XXX/connectors/XXX/devices?page[size]=1&page[before]=Mzk5ZTM1MWYtNTI3OS00YzFhLTk0MmUtYTZiODBmMjFiYzVh";

        params = api.Util.parseURLParameter(url)

        self.assertEqual(int(params["page[size]"]), 1, "Page[size] should be 1")
        self.assertEqual(params["page[before]"], "Mzk5ZTM1MWYtNTI3OS00YzFhLTk0MmUtYTZiODBmMjFiYzVh", "page[before] should be Mzk5ZTM1MWYtNTI3OS00YzFhLTk0MmUtYTZiODBmMjFiYzVh")


class TestParseCurrentDateTime(unittest.TestCase):

    def test_parsing(self):
            
        currentDateTime = int(time.time())

        print("CURRENT DATETIME: ", currentDateTime)

        currentDateTimeString = datetime.datetime.fromtimestamp(currentDateTime).strftime("%FT%T.000Z")

        print("CURRENT DATETIME STR: ", currentDateTimeString)

        parsedDateTime = api.Util.parseDateTimeString(currentDateTimeString)
        parsedDateTimeString = datetime.datetime.fromtimestamp(parsedDateTime).strftime("%FT%T.000Z")

        print("PARSED DATETIME: ", parsedDateTime)
        print("PARSED DATETIME STR: ", parsedDateTimeString)

        self.assertEqual(currentDateTime, parsedDateTime, "Timestamps should be equal")
        self.assertEqual(currentDateTimeString, parsedDateTimeString, "Dates should be equal")


class TestParseFixedDateTime(unittest.TestCase):

    def test_parsing(self):
            
        fixedDateTimeString = "2019-06-02T12:17:58.819Z"

        print("FIXED DATETIME STR: ", fixedDateTimeString)

        parsedDateTime = api.Util.parseDateTimeString(fixedDateTimeString)
        parsedDateTimeString = datetime.datetime.fromtimestamp(parsedDateTime).strftime("%FT%T.819Z")

        print("PARSED DATETIME: ", parsedDateTime)
        print("PARSED DATETIME STR: ", parsedDateTimeString)

        self.assertEqual(fixedDateTimeString, parsedDateTimeString, "Dates should be equal")

if __name__ == '__main__':
    unittest.main()